Caused by: java.lang.NoSuchFieldException: jvm

I want to profile my gradle build with the following commandline: `gradle-profiler --benchmark --profile async-profiler --project-dir ./  compileTestJava`

I get the following exception during the run:

```
* Stopping daemons
org.gradle.tooling.BuildException: Could not execute build using connection to Gradle installation '/project/gradle-user-home/wrapper/dists/gradle-6.5.1-bin/1m5048aptkfynhbvolwgr4ej9/gradle-6.5.1'.
	at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:51)
	at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:29)
	at org.gradle.tooling.internal.consumer.ResultHandlerAdapter.onFailure(ResultHandlerAdapter.java:43)
	at org.gradle.tooling.internal.consumer.async.DefaultAsyncConsumerActionExecutor.lambda$run$0(DefaultAsyncConsumerActionExecutor.java:64)
	at org.gradle.tooling.internal.consumer.async.DefaultAsyncConsumerActionExecutor$$Lambda$29/000000000000000000.run(Unknown Source)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.base/java.lang.Thread.run(Thread.java:836)
	at org.gradle.tooling.internal.consumer.BlockingResultHandler.getResult(BlockingResultHandler.java:46)
	at org.gradle.tooling.internal.consumer.DefaultBuildLauncher.run(DefaultBuildLauncher.java:83)
	at org.gradle.profiler.ToolingApiGradleClient.lambda$runTasks$0(ToolingApiGradleClient.java:51)
	at org.gradle.profiler.ToolingApiGradleClient$$Lambda$69/000000000000000000.apply(Unknown Source)
	at org.gradle.profiler.ToolingApiGradleClient.run(ToolingApiGradleClient.java:29)
	at org.gradle.profiler.ToolingApiGradleClient.runOperation(ToolingApiGradleClient.java:42)
	at org.gradle.profiler.ToolingApiGradleClient.runTasks(ToolingApiGradleClient.java:47)
	at org.gradle.profiler.RunTasksAction.run(RunTasksAction.java:36)
	at org.gradle.profiler.GradleInvokerBuildAction.run(GradleInvokerBuildAction.java:10)
	at org.gradle.profiler.BuildUnderTestInvoker$InvokeAndMeasureAction.run(BuildUnderTestInvoker.java:59)
	at org.gradle.profiler.BuildUnderTestInvoker$InvokeAndMeasureAction.run(BuildUnderTestInvoker.java:40)
	at org.gradle.profiler.RunBuildStepAction.run(RunBuildStepAction.java:29)
	at org.gradle.profiler.ScenarioInvoker.runMeasured(ScenarioInvoker.java:36)
	at org.gradle.profiler.GradleScenarioInvoker.doRun(GradleScenarioInvoker.java:113)
	at org.gradle.profiler.GradleScenarioInvoker.doRun(GradleScenarioInvoker.java:21)
	at org.gradle.profiler.ScenarioInvoker.run(ScenarioInvoker.java:17)
	at org.gradle.profiler.Main.invoke(Main.java:110)
	at org.gradle.profiler.Main.run(Main.java:76)
	at org.gradle.profiler.Main.main(Main.java:18)
Caused by: org.gradle.internal.exceptions.LocationAwareException: Failed to create service 'pidCollector'.
	at org.gradle.initialization.exception.DefaultExceptionAnalyser.transform(DefaultExceptionAnalyser.java:103)
	at org.gradle.initialization.exception.DefaultExceptionAnalyser.collectFailures(DefaultExceptionAnalyser.java:67)
	at org.gradle.initialization.exception.MultipleBuildFailuresExceptionAnalyser.transform(MultipleBuildFailuresExceptionAnalyser.java:40)
	at org.gradle.initialization.exception.StackTraceSanitizingExceptionAnalyser.transform(StackTraceSanitizingExceptionAnalyser.java:30)
	at org.gradle.initialization.DefaultGradleLauncher.finishBuild(DefaultGradleLauncher.java:201)
	at org.gradle.initialization.DefaultGradleLauncher.finishBuild(DefaultGradleLauncher.java:127)
	at org.gradle.internal.invocation.GradleBuildController$1.create(GradleBuildController.java:75)
	at org.gradle.internal.invocation.GradleBuildController$1.create(GradleBuildController.java:67)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:180)
	at org.gradle.internal.work.StopShieldingWorkerLeaseService.withLocks(StopShieldingWorkerLeaseService.java:40)
	at org.gradle.internal.invocation.GradleBuildController.doBuild(GradleBuildController.java:67)
	at org.gradle.internal.invocation.GradleBuildController.run(GradleBuildController.java:56)
	at org.gradle.tooling.internal.provider.runner.BuildModelActionRunner.run(BuildModelActionRunner.java:54)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
	at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:63)
	at org.gradle.tooling.internal.provider.ValidatingBuildActionRunner.run(ValidatingBuildActionRunner.java:32)
	at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:39)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:51)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner$3.call(RunAsBuildOperationBuildActionRunner.java:45)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:409)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:399)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:94)
	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
	at org.gradle.launcher.exec.RunAsBuildOperationBuildActionRunner.run(RunAsBuildOperationBuildActionRunner.java:45)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:50)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter$1.transform(InProcessBuildActionExecuter.java:47)
	at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:80)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:47)
	at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:31)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:42)
	at org.gradle.launcher.exec.BuildTreeScopeBuildActionExecuter.execute(BuildTreeScopeBuildActionExecuter.java:28)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:87)
	at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:55)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:60)
	at org.gradle.tooling.internal.provider.SubscribableBuildActionExecuter.execute(SubscribableBuildActionExecuter.java:38)
	at org.gradle.tooling.internal.provider.SessionScopeBuildActionExecuter.execute(SessionScopeBuildActionExecuter.java:68)
	at org.gradle.tooling.internal.provider.SessionScopeBuildActionExecuter.execute(SessionScopeBuildActionExecuter.java:38)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:37)
	at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:26)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:60)
	at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:32)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:56)
	at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:42)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:48)
	at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:32)
	at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:68)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:78)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:75)
	at org.gradle.util.Swapper.swap(Swapper.java:38)
	at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:75)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:82)
	at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
	at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
	at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
	at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
Caused by: org.gradle.api.services.internal.ServiceLifecycleException: Failed to create service 'pidCollector'.
	at org.gradle.api.services.internal.BuildServiceProvider.calculateOwnValue(BuildServiceProvider.java:106)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:84)
	at org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry$ForwardingBuildEventConsumer.handle(DefaultBuildEventsListenerRegistry.java:234)
	at org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry$ForwardingBuildEventConsumer.handle(DefaultBuildEventsListenerRegistry.java:211)
	at org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry$AbstractListener.run(DefaultBuildEventsListenerRegistry.java:153)
	at org.gradle.internal.build.event.DefaultBuildEventsListenerRegistry$AbstractListener$$Lambda$244/000000000000000000.run(Unknown Source)
	... 3 more
Caused by: org.gradle.api.reflect.ObjectInstantiationException: Could not create an instance of type org.gradle.trace.pid.PidCollectorBuildService.
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:69)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:55)
	at org.gradle.api.services.internal.BuildServiceProvider.calculateOwnValue(BuildServiceProvider.java:104)
	... 8 more
Caused by: java.lang.NoSuchFieldException: jvm
	at org.gradle.trace.pid.PidCollectorBuildService.getPid(PidCollectorBuildService.java:39)
	at org.gradle.trace.pid.PidCollectorBuildService.<init>(PidCollectorBuildService.java:32)
	at org.gradle.trace.pid.PidCollectorBuildService$Inject.<init>(Unknown Source)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at org.gradle.internal.instantiation.generator.AsmBackedClassGenerator$InvokeConstructorStrategy.newInstance(AsmBackedClassGenerator.java:1969)
	at org.gradle.internal.instantiation.generator.AbstractClassGenerator$GeneratedClassImpl$GeneratedConstructorImpl.newInstance(AbstractClassGenerator.java:468)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:64)
```

gradle: 6.5.1
java: openjdk 11.0.9 2020-10-20 Eclipse OpenJ9 VM AdoptOpenJDK (build openj9-0.23.0, JRE 11 Linux amd64-64-Bit Compressed References 20201022_810 (JIT enabled, AOT enabled)

--- a/subprojects/build-operations/build.gradle
+++ b/subprojects/build-operations/build.gradle
@@ -1,7 +1,0 @@
-plugins {
-    id 'profiler.embedded-library'
-}
-
-dependencies {
-    api gradleApi()
-}

--- a/subprojects/build-operations/build.gradle.kts
+++ b/subprojects/build-operations/build.gradle.kts
@@ -1,0 +1,18 @@
+plugins {
+    id("profiler.embedded-library")
+}
+
+dependencies {
+    api(gradleApi())
+}
+
+java {
+    toolchain {
+        languageVersion.set(JavaLanguageVersion.of(11))
+    }
+}
+
+tasks.withType<AbstractCompile>().configureEach {
+    targetCompatibility = "8"
+    sourceCompatibility = "8"
+}

--- a/subprojects/build-operations/src/main/java/org/gradle/trace/pid/AbstractPidCollectorBuildService.java
+++ b/subprojects/build-operations/src/main/java/org/gradle/trace/pid/AbstractPidCollectorBuildService.java
@@ -1,0 +1,37 @@
+package org.gradle.trace.pid;
+
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.internal.GradleInternal;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.build.event.BuildEventsListenerRegistry;
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.OperationCompletionListener;
+import org.gradle.util.GFileUtils;
+
+import java.io.File;
+
+public abstract class AbstractPidCollectorBuildService implements BuildService<AbstractPidCollectorBuildService.Parameters>, OperationCompletionListener {
+
+    public static <T extends AbstractPidCollectorBuildService> void registerBuildService(Class<T> buildServiceClass, GradleInternal gradle, File outFile) {
+        Provider<T> pidCollectorService = gradle.getSharedServices().registerIfAbsent("pidCollector", buildServiceClass, spec -> spec.parameters(
+            params -> params.getPidFile().set(outFile)
+        ));
+        gradle.getServices().get(BuildEventsListenerRegistry.class).onTaskCompletion(pidCollectorService);
+    }
+
+    public interface Parameters extends BuildServiceParameters {
+        RegularFileProperty getPidFile();
+    }
+
+    protected AbstractPidCollectorBuildService() {
+        GFileUtils.writeFile(getPid().toString(), getParameters().getPidFile().get().getAsFile());
+    }
+
+    protected abstract Long getPid();
+
+    @Override
+    public void onFinish(FinishEvent event) {
+    }
+}

--- a/subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollector.java
+++ b/subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollector.java
@@ -1,5 +1,6 @@
 package org.gradle.trace.pid;
 
+import org.gradle.api.JavaVersion;
 import org.gradle.api.internal.GradleInternal;
 import org.gradle.internal.nativeintegration.ProcessEnvironment;
 import org.gradle.util.GFileUtils;
@@ -14,7 +15,11 @@ public class PidCollector {
         ProcessEnvironment processEnvironment = gradle.getServices().get(ProcessEnvironment.class);
         GFileUtils.writeFile(processEnvironment.getPid().toString(), outFile);
         if (GradleVersion.current().compareTo(GradleVersion.version("6.1")) >= 0) {
-            PidCollectorBuildService.registerBuildService(gradle, outFile);
+            if (JavaVersion.current().isJava9Compatible()) {
+                AbstractPidCollectorBuildService.registerBuildService(PidCollectorBuildService.class, gradle, outFile);
+            } else {
+                AbstractPidCollectorBuildService.registerBuildService(PidCollectorJava8BuildService.class, gradle, outFile);
+            }
         }
     }
 }

--- a/subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollectorBuildService.java
+++ b/subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollectorBuildService.java
@@ -1,53 +1,9 @@
 package org.gradle.trace.pid;
 
-import org.gradle.api.file.RegularFileProperty;
-import org.gradle.api.internal.GradleInternal;
-import org.gradle.api.provider.Provider;
-import org.gradle.api.services.BuildService;
-import org.gradle.api.services.BuildServiceParameters;
-import org.gradle.build.event.BuildEventsListenerRegistry;
-import org.gradle.tooling.events.FinishEvent;
-import org.gradle.tooling.events.OperationCompletionListener;
-import org.gradle.util.GFileUtils;
-
-import java.io.File;
-import java.lang.management.ManagementFactory;
-import java.lang.management.RuntimeMXBean;
-import java.lang.reflect.InvocationTargetException;
-
-public abstract class PidCollectorBuildService implements BuildService<PidCollectorBuildService.Parameters>, OperationCompletionListener {
-
-    public static void registerBuildService(GradleInternal gradle, File outFile) {
-        Provider<PidCollectorBuildService> pidCollectorService = gradle.getSharedServices().registerIfAbsent("pidCollector", PidCollectorBuildService.class, spec -> spec.parameters(
-            params -> params.getPidFile().set(outFile)
-        ));
-        gradle.getServices().get(BuildEventsListenerRegistry.class).onTaskCompletion(pidCollectorService);
-    }
-
-    public interface Parameters extends BuildServiceParameters {
-        RegularFileProperty getPidFile();
-    }
-
-    public PidCollectorBuildService() throws Exception {
-        GFileUtils.writeFile(getPid().toString(), getParameters().getPidFile().get().getAsFile());
-    }
-
-    private Integer getPid() throws NoSuchFieldException, IllegalAccessException, NoSuchMethodException, InvocationTargetException {
-        // With Gradle 6.6 we can't inject `ProcessEnvironment`, so that is why we access the runtime MXBean here.
-        // On Java 9 we could use the nicer ProcessHandle.
-        RuntimeMXBean runtime = ManagementFactory.getRuntimeMXBean();
-        java.lang.reflect.Field jvm = runtime.getClass().getDeclaredField("jvm");
-        jvm.setAccessible(true);
-        sun.management.VMManagement mgmt =
-            (sun.management.VMManagement) jvm.get(runtime);
-        java.lang.reflect.Method pid_method =
-            mgmt.getClass().getDeclaredMethod("getProcessId");
-        pid_method.setAccessible(true);
-
-        return (Integer) pid_method.invoke(mgmt);
-    }
+public abstract class PidCollectorBuildService extends AbstractPidCollectorBuildService {
 
     @Override
-    public void onFinish(FinishEvent event) {
+    protected Long getPid() {
+        return ProcessHandle.current().pid();
     }
 }

--- a/subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollectorJava8BuildService.java
+++ b/subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollectorJava8BuildService.java
@@ -1,0 +1,28 @@
+package org.gradle.trace.pid;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.RuntimeMXBean;
+import java.lang.reflect.InvocationTargetException;
+
+public abstract class PidCollectorJava8BuildService extends AbstractPidCollectorBuildService {
+
+    @Override
+    protected Long getPid() {
+        try {
+            // With Gradle 6.6 we can't inject `ProcessEnvironment`, so that is why we access the runtime MXBean here.
+            // On Java 9 we could use the nicer ProcessHandle.
+            RuntimeMXBean runtime = ManagementFactory.getRuntimeMXBean();
+            java.lang.reflect.Field jvm = runtime.getClass().getDeclaredField("jvm");
+            jvm.setAccessible(true);
+            sun.management.VMManagement mgmt =
+                (sun.management.VMManagement) jvm.get(runtime);
+            java.lang.reflect.Method pid_method =
+                mgmt.getClass().getDeclaredMethod("getProcessId");
+            pid_method.setAccessible(true);
+
+            return ((Integer) pid_method.invoke(mgmt)).longValue();
+        } catch (NoSuchFieldException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

--- a/subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollectorJava8BuildService.java
+++ b/subprojects/build-operations/src/main/java/org/gradle/trace/pid/PidCollectorJava8BuildService.java
@@ -10,7 +10,7 @@ public abstract class PidCollectorJava8BuildService extends AbstractPidCollector
     protected Long getPid() {
         try {
             // With Gradle 6.6 we can't inject `ProcessEnvironment`, so that is why we access the runtime MXBean here.
-            // On Java 9 we could use the nicer ProcessHandle.
+            // On Java 9 we use `ProcessHandle` via `PidCollectorBuildService`.
             RuntimeMXBean runtime = ManagementFactory.getRuntimeMXBean();
             java.lang.reflect.Field jvm = runtime.getClass().getDeclaredField("jvm");
             jvm.setAccessible(true);

